File: Q6/cpp/FriendScore.cpp
#include <vector>
#include <string>
#include <iostream>
#include <algorithm>

using namespace std;

class FriendScore
{
public:
    //중복을 피하기 위해서 미리 한번씩 다계산하여 최종적으로 카운팅함.
    int highestScore(vector<string> friends)
    {
        vector<int> friendsNum = vector<int>(friends.size(),0);
        for(int i = 0;i<friends.size();i++)
        {
            vector<int> commonFriends;
            for(int j = 0; j<friends[i].size();j++)
            {
                char y = 'Y';
                if(friends[i].at(j) == y)
                {
                    //friendsNum[i] ++;
                    commonFriends.push_back(j);
                }
            }

            for(int n = 0; n<commonFriends.size();n++)
            {
                for(int m = 0; m<commonFriends.size();m++)
                {
                    if(n!=m)
                    {
                        friends[commonFriends[n]].at(commonFriends[m]) = 'Y';
                    }
                }
            }
        }

        cout<<endl;

        for(int i = 0;i<friends.size();i++)
        {
            for(int j = 0; j<friends[i].size();j++)
            {
                char y = 'Y';
                if(friends[i].at(j) == y)
                {
                    friendsNum[i] ++;
                }
                cout<<friends[i].at(j)<<" ";
            }
            cout<<endl;
        }
        
        int maxNum = 0;
        vector<int>::iterator it;
        for(it = friendsNum.begin() ; it != friendsNum.end(); it++)
        {
            maxNum = max(maxNum,*it);
        }
        return maxNum;
    }
};

int main()
{
    vector<string> friends_1 = {
                                "NNN",
                                "NNN",
                                "NNN"
                             };

    vector<string> friends_2 = {
                                "NYY",
                                "YNY",
                                "YYN"
                             };

    vector<string> friends_3 = {
                                "NYNNN",
                                "YNYNN",
                                "NYNYN",
                                "NNYNY",
                                "NNNYN"
                             };

    vector<string> friends_4 = {
                                "NNNNNNNNNNNNNNY",
                                "NNNNNNNNNNNNNNN",
                                "NNNNNNYNNNNNNNN",
                                "NNNNNNYNNNNNNNY",
                                "NNNNNNNNNNNNNNY",
                                "NNNNNNNNYNNNNNN",
                                "NNNNNNNNNNNNNNN",
                                "NNYYNNNNNNNNNNN",
                                "NNNNNYNNNNNYNNN",
                                "NNNNNNNNNNNNNNY",
                                "NNNNNNNNNNNNNNN",
                                "NNNNNNNNYNNNNNN",
                                "NNNNNNNNNNNNNNN",
                                "NNNNNNNNNNNNNNN",
                                "YNNYYNNNNYNNNNN"
                             };
                             
    cout<<FriendScore().highestScore(friends_1)<<endl;
    cout<<FriendScore().highestScore(friends_2)<<endl;
    cout<<FriendScore().highestScore(friends_3)<<endl;
    cout<<FriendScore().highestScore(friends_4)<<endl;
    return 0;
}

File: Q6/python/FriendScore.py
class FriendScore:
    def __init__(self):
        pass

    def highestScore(self, friends : list):
        friendlist = friends
        #mutable 이기 때문에 callbyreference가 되버림. 계산과정을 기록하는건 원치않으므로 리스트를 새로작성.
        friendnum =  [0] * friendlist.__len__()
        for i in range(0,friendlist.__len__()):
            commonFriends = list()
            for j in range(0,friendlist.__len__()):
                if friendlist[i][j] == 'Y':
                    commonFriends.append(j)

            for n in range(0,commonFriends.__len__()):
                for m in range(0,commonFriends.__len__()):
                    if n != m:
                        #str은 immutable 형태이기때문에 list형태로 변환뒤 조작후 다시 str형태로 변환
                        changeStr : str= friendlist[commonFriends[n]]
                        changeList = []
                        for c in range(0,changeStr.__len__()):
                            changeList.append(changeStr[c])
                        changeList[commonFriends[m]] = 'Y'

                        sumStr = ''.join(changeList)
                        friendlist[commonFriends[n]] = sumStr

        #위에서 사용했으므로 한번 초기화
        i = 0
        j = 0
        for i in range(0,friendlist.__len__()):
            for j in range(0,friendlist.__len__()):
                if friendlist[i][j] == 'Y':
                    friendnum[i]+= 1
                print(friendlist[i][j] + " ",end="")
            print("")

        friendnum.sort()
        return friendnum[-1]

friends_1 = [
                "NNN",
                "NNN",
                "NNN"
             ]

friends_2 = [
                "NYY",
                "YNY",
                "YYN"
             ]

friends_3 = [
                "NYNNN",
                "YNYNN",
                "NYNYN",
                "NNYNY",
                "NNNYN"
             ]

friends_4 = [
            "NNNNNNNNNNNNNNY",
            "NNNNNNNNNNNNNNN",
            "NNNNNNYNNNNNNNN",
            "NNNNNNYNNNNNNNY",
            "NNNNNNNNNNNNNNY",
            "NNNNNNNNYNNNNNN",
            "NNNNNNNNNNNNNNN",
            "NNYYNNNNNNNNNNN",
            "NNNNNYNNNNNYNNN",
            "NNNNNNNNNNNNNNY",
            "NNNNNNNNNNNNNNN",
            "NNNNNNNNYNNNNNN",
            "NNNNNNNNNNNNNNN",
            "NNNNNNNNNNNNNNN",
            "YNNYYNNNNYNNNNN"
         ]

print(FriendScore().highestScore(friends_1))
print(FriendScore().highestScore(friends_2))
print(FriendScore().highestScore(friends_3))
print(FriendScore().highestScore(friends_4))

